### ISSUE TITLE
FAISS + Windows

### ISSUE THREAD

**Issue by Sterilise** [KevinMusgrave/pytorch-metric-learning#251]
stat_util.py uses the FAISS library, this library doesn't have any prebuilt versions for windows, due to this I cannot use any testers offered by this library. I'm not a very technical person (when it comes to building c++ libraries), is there an easy way to circumvent this library?

**Comment by KevinMusgrave** [comment]
I'm not sure how to get faiss working on windows.

Probably the best solution will be to change the AccuracyCalculator class so that it allows a k-nearest-neighbor and k-means function to be passed in. That way if you can't use faiss, you can use some other library. I'll try to get this done for the next release.

**Comment by KevinMusgrave** [comment]
In v1.0.0 the knn function can be customized, so you can use something compatible with windows instead of faiss. See docs: https://kevinmusgrave.github.io/pytorch-metric-learning/accuracy_calculation/

**Comment by seanyeo300** [comment]
Hi Kevin, I would like to know if I can run FAISS using the CPU version while training using GPU? I am not too sure how to modify the KNN and K-means function to be passed into the accuracy calculator like what you mentioned above.

**Comment by KevinMusgrave** [comment]
Yeah if I remember correctly, `AccuracyCalculator` takes in a `device` argument, so you can set the device to be `torch.device("cpu")`

**Comment by seanyeo300** [comment]
Thanks! Just another question, I am trying to use the contrastive loss with L1 distance instead of Euclidean distance. I was looking at the documentation but can't find a way to control the type of norm used.

**Comment by seanyeo300** [comment]
Thank you!
